aws/ecs_cluster.go

```diff
@@ -0,0 +1,132 @@
+package aws
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+const firstSeenTagKey = "cloud-nuke-first-seen"
+
+// Tag an ECS cluster identified by the given cluster ARN with a tag that has the given key and value
+func tagEcsCluster(awsSession *session.Session, clusterArn *string, tagKey string, tagValue string) error {
+	svc := ecs.New(awsSession)
+	input := &ecs.TagResourceInput{
+		ResourceArn: clusterArn,
+		Tags: []*ecs.Tag{
+			{
+				Key:   aws.String(tagKey),
+				Value: aws.String(tagValue),
+			},
+		},
+	}
+
+	_, err := svc.TagResource(input)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	return nil
+}
+
+func getAllEcsClustersOlderThan(awsSession *session.Session, region string, excludeAfter time.Time) ([]*string, error) {
+	awsSession, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region),
+	})
+
+	clusterArns, err := getAllEcsClusters(awsSession)
+	if err != nil {
+		logging.Logger.Errorf("Error getting all ECS clusters")
+		return nil, err
+	}
+
+	var filteredEcsClusters []*string
+	for _, clusterArn := range clusterArns {
+
+		firstSeenTime, err := getClusterTag(awsSession, clusterArn, firstSeenTagKey)
+		if err != nil {
+			logging.Logger.Errorf("Error getting the `first_seen` tag for ECS cluster with ARN %s", *clusterArn)
+			return nil, err
+		}
+
+		if firstSeenTime.IsZero() {
+			err := tagEcsCluster(awsSession, clusterArn, firstSeenTagKey, time.Now().UTC().String())
+			if err != nil {
+				logging.Logger.Errorf("Error tagigng the ECS cluster with ARN %s", *clusterArn)
+				return nil, err
+			}
+		}
+
+		if excludeAfter.After(firstSeenTime) {
+			filteredEcsClusters = append(filteredEcsClusters, clusterArn)
+		}
+	}
+
+	return filteredEcsClusters, nil
+}
+
+func nukeEcsClusters(awsSession *session.Session, ecsClusterArns []*string) error {
+	svc := ecs.New(awsSession)
+
+	numNuking := len(ecsClusterArns)
+
+	if numNuking == 0 {
+		logging.Logger.Infof("No ECS clusters to nuke in region %s", *awsSession.Config.Region)
+		return nil
+	}
+
+	logging.Logger.Infof("Deleting %d ECS clusters in region %s", numNuking, *awsSession.Config.Region)
```

Comment:
Use `aws.StringValue` to more safely dereference pointers in the AWS SDK.

Reply:
Thanks! I mistakenly hard-resetted 😱 that yesterday and forgot to re-add. I have now updated all pointers to be `aws.StringValue()`. Could you point me to some guidance why it's better than just using the pointer? I found this so far, which is somewhat explanatory: https://github.com/aws/aws-sdk-go/issues/114

Reply:
The AWS Go SDK uses pointers all over the place as a way to handle optional values: that is, data that might or might not be in the JSON/XML data that's exchanged during API calls. That means that any of those pointers could be `nil`. If you dereference a `nil` pointer in Go, you get a panic, and your app exits, with no chance for clean up code or anything else to run. You can add things to your code to catch (or more accurately, "recover from") panics, but it's a messy andd error prone business. You can also read the docs and try to deeply understand exactly which values are guaranteed to be non-`nil`, but again, it's a messy and error prone business (and may change as the API evolves).

The `XXXValue` methods (e.g., `StringValue`) in the SDK do a `nil`-safe dereference: e.g., if the value is `nil`, `StringValue` returns an empty string.

So, in general, we:

1. Never use `panic` ourselves in code. Instead, we always return an `error` all the way up the stack, and only the main entry method may choose to `panic` or exit in some other way.
1. Handle every single `error` that is returned. We never ignore them.
1. Use `XXXValue` methods whenever derferencing `nil` pointers in the AWS Go SDK.
1. Check for `nil` and empty arrays whenever possible.

Reply:
I just want to say that I really appreciate in-depth and well thought-out answers like this. As a new contributor, it helps _so_ much to ramp up more quickly and easily. Thank you, @brikis98!

Reply:
Glad to hear it!

Actually, something @yorinasub17 started recently that I think is great is creating style guides. We have one here for Terraform: https://www.notion.so/gruntwork/Style-guides-0756577007fc41d9916ad34e9940d202. Perhaps we should start building a page for Go? For the most part, we follow Go's official conventions (incl what's enforced by `go fmt`), but some of the items above our uniquely our own. I'm a bit buried today, but if someone wants to start that page on our wiki, copy some of the above, and what's in this comment, I'd be happy to contribute (and I'm sure many others at the company would too).

Reply:
A style guide for Go is a fantastic idea. I'd love to start this. Will post a comment back here once I have something up.

Reply:
This is very helpful @brikis98 Thank you! Will keep this in mind in future :)